widget/entry-with-completion.go

```diff
@@ -0,0 +1,127 @@
+package widget
+
+import (
+	"fyne.io/fyne/v2"
+	"fyne.io/fyne/v2/widget"
+)
+
+// CompletionEntry is an Entry with options displayed in a PopUpMenu.
+type CompletionEntry struct {
+	widget.Entry
+	menu      *fyne.Menu
+	popupMenu *widget.PopUpMenu
+	items     []string
+	pause     bool
+}
+
+// NewCompletionEntry creates a new CompletionEntry which creates a popup menu that responds to keystrokes to navigate through the items without losing the editing ability of the text input..
+func NewCompletionEntry(options []string) *CompletionEntry {
+	c := &CompletionEntry{items: options}
+	c.ExtendBaseWidget(c)
+	// c.Entry = widget.NewEntry()
+	return c
+}
+
+// ShowCompletion displays the completion menu
+func (c *CompletionEntry) ShowCompletion() {
+	if c.pause {
+		return
+	}
+	items := make([]*fyne.MenuItem, len(c.items))
+	for i, item := range c.items {
+		s := item // dereferencing
+		items[i] = fyne.NewMenuItem(s, func() {
+			c.setTextFromMenu(s)
+			c.popupMenu.Hide()
+		})
+	}
+	if c.popupMenu != nil {
+		c.popupMenu.Hide()
+	}
+	// New menu
+	c.menu = fyne.NewMenu("", items...)
+	holder := fyne.CurrentApp().Driver().CanvasForObject(c)
+	c.popupMenu = widget.NewPopUpMenu(c.menu, holder)
+	c.bindKey(holder)
+
+	pos := fyne.CurrentApp().Driver().AbsolutePositionForObject(c)
+	c.popupMenu.ShowAtPosition(fyne.Position{X: pos.X, Y: pos.Y + c.Size().Height})
+}
+
+// HideCompletion hides the completion menu.
+func (c *CompletionEntry) HideCompletion() {
+	if c.popupMenu != nil {
+		c.popupMenu.Hide()
+	}
+}
+
+// SetOptions set the completion list with itemList.
+func (c *CompletionEntry) SetOptions(itemList []string) {
+	c.items = itemList
+}
+
+// Prevent the menu to open when the user validate value from the menu.
+func (c *CompletionEntry) setTextFromMenu(s string) {
+	c.pause = true
+	c.Entry.SetText(s)
+	c.Entry.Refresh()
+	c.pause = false
+}
+
+// FocusGained make the Entry to be focused.
+//
+// Implements: fyne.Focusable
+func (c *CompletionEntry) FocusGained() {
+	c.Entry.FocusGained()
+}
+
+// FocusLose is only here to respect interface.
+//
+// Implements: fyne.Focusable
+func (c *CompletionEntry) FocusLose() {}
+
+// TypedRune sends the rune to the Entry, whatever if the popup menu for completion is displayed or not.
+//
+// Implements: fyne.Focusable
+func (c *CompletionEntry) TypedRune(r rune) {
+	c.Entry.TypedRune(r)
+}
+
+// TypedKey sends the key events to the Entry, whatever if the popup menu for completion is displayed or not.
+//
+// Implements: fyne.Focusable
+func (c *CompletionEntry) TypedKey(event *fyne.KeyEvent) {
+	c.Entry.TypedKey(event)
+}
+
+// Bind event and rune key to the Entry if needed, and bind othters navigation key + validation with Return.
+// This bound keys are set when the "popupmenu" gains the focus, the canvas continues to send rune and keys
+// to the Entry.
+func (c *CompletionEntry) bindKey(cnv fyne.Canvas) {
+	position := -1
+	cnv.SetOnTypedRune(func(r rune) {
```

Comment:
I am unsure about this. Canvas callbacks could be edited at application level by the developer. So by doing this, there would be a chance that the developer change this in its application (then the completion entry would not work), or viceversa, the completion entry would override the developer setting.

Reply:
The entry loses the focus while the popup menu is displayed, there is no other way to manage navigation for popup menu **and** continue to edit the entry without having key bound at canvas levelr :( I spent hours to find solutions, speaking with Andrew and trying maybe 10 implementations. This is the only one that works correctly :)

Reply:
Yes, I know the reason behind this, but unfortunately this would lead to that kind of problems :(

Reply:
In my opinion, it's a shame to prevent yourself from having this type of widget simply because the basic library suffers from some design issues.
We would be depriving ourselves of offering some very useful tools, really very useful, because Fyne doesn't offer some routines.
Here the widget works, potentially we could add a note that this widget cannot be overloaded at the canvas level.
But this widget is one of the first ones that I complained about that it doesn't exist... And I admit that since version 1 of Fyne, I don't expect a major API change that allows us to have completion on an Entry...

Reply:
(sorry, I used the word "shame" because I was helped by a translator... and didn't seen that word. I would say "a pity", I fixed it)

Reply:
I am with you, I really like this new widget :). But maybe we should find a solution for this first. For example, canvas could add listeners to that events instead of set (replace) them.

Reply:
It's also possible to mark this as "experimental" until something changes in Fyne to allow me to not use the Canvas to bind keys... But really, this widget is a big miss in Fyne, and I though that fyne-x could propose it ;)

Reply:
"But maybe we should find a solution for this first. For example, canvas could add listeners to that events instead of set (replace) them."
I don't see what you mean. Is this not what does the "TypedKey" and "TypedRune" ?

Reply:
> I don't see what you mean. Is this not what does the "TypedKey" and "TypedRune" ?

I mean this `canvas.SetOnTypedRune` could add a functionality instead of replace it.

By the way, I think this could be possible to do without using canvas bindings, maybe a custom popup menu that listen to typed runes or keys?

Reply:
That's what I tried before. The problem is not the popupmenu but that it is in a secondary canvas (overlay). If I do not do this, I cannot place the menu because I don't have the container where to place the "menu".
SetOnTypedRune failed for me, I lose the cursor in the Entry (and I don't know why).
If you've got a begining of what I could change (because I really don't see other way right now), that could be nice. I will take a look.

Reply:
Ok I have done some tests, and it is possible to do without canvas bindings :). However, currently is not possible to do with the current version of Fyne. Because this requires to be able to focus manually objects in an overlay which I think it is not possible to do outside the fyne repository (internal implementation). But that should be added soon I think :)

Reply:
>  The problem is not the popupmenu but that it is in a secondary canvas (overlay). 

Absolutely. There should be a function to focus overlay content objects. Sorry, I wrote the above comment, without realized that you had answered.

Reply:
> Because this requires to be able to focus manually objects in an overlay

this should be possible. The code changed on `develop` but it should have worked before as well.

I agree that a component such as this needs to use focused based key events instead of canvas hooks. If it is not working/possible then we have a bug that should be fixed

Reply:
Introducing that function, would allow us to do this without canvas bindings. However this time, you should implement manually the menu (that should implement `fyne.Focusable`) with a widget.PopUp (not a PopUpMenu).

https://user-images.githubusercontent.com/12239342/115155041-9bad2200-a043-11eb-86b1-5c2f5fd48d4e.mp4

Reply:
But anyway I will need to put it in the overlay, are we OK ?

Reply:
@andydotxyz 
> If it is not working/possible then we have a bug that should be fixed

Yes, this should be fixed, basically the problem is that each overlay has its own focus manager.

> But anyway I will need to put it in the overlay, are we OK ?

Yes, absolutely. You should create the menu manually and then pass it to `widget.NewPopUp(menu, canvas)`.

Reply:
OK, I'll do it in a couple of hours, thanks for your time and idea. I hope to propose this to the community, Fyne and Fyne-X are really cool stuffs :)

Reply:
> thanks for your time and idea

On the contrary thanks to you, this is a great addition :)

Reply:
@metal3d
Have just realized that it is indeed possible to do it currently, as @andydotxyz said.

When you already build the menu manually and insert it to the widget.PopUp, you can focus the menu immediately after it is shown by doing:
```go
c.popup = widget.NewPopUp(menu, holder)
c.popup.Resize(fyne.NewSize(c.Entry.Size().Width, c.popup.MinSize().Height))

pos := fyne.CurrentApp().Driver().AbsolutePositionForObject(c)
c.popup.ShowAtPosition(fyne.Position{X: pos.X, Y: pos.Y + c.Size().Height})
holder.Focus(menu) // 👈 added
```
Note that `holder.Focus` needs to be called just after `popup.ShowAtPosition(...)` because the overlay only exist when it is shown.

Reply:
This is not the wanted behavior because the cursor disappears from the entry

Reply:
@metal3d Did you try it? The cursor will not disappear because overlays have their own focus scope, so by doing `holder.Focus(menu)`, the entry will remain focused.

Reply:
I tried.
There is a problem with `widget.Menu` (overriden in a popup) that won't take the size of the popup and cannot be "updated", so I'm now using a List (with data bind). 
I implemented the Focusable interface, I print a log in `FocusGained`.

With a simple example: no problem (there is a problem with the List height but let's check the basics)

```go
func (n *navigableList) FocusGained() {
	log.Println("Focused !")
}
```

In test:
```go
	entry := createEntry()
	win := test.NewWindow(entry)
	win.Resize(fyne.NewSize(500, 300))
	defer win.Close()

	entry.SetText("init") // here the "Focused !" log appears from FocusGained
	entry.ShowCompletion() // => displays <nil>
	elem := test.Canvas().Overlays().Top()
	log.Println("Focused => ", elem)
```

```
=== RUN   TestCompletionEntry_ShowMenu
2021/04/19 10:32:43 Focused !
2021/04/19 10:32:43 Focused =>  <nil>
```

I tried to get `win.Canvas().Focus()` instead of Overlay => same, it's "<nil>".

Whatever I try, it works on app, but the test won't give me the focused object.

I also tried:

```go
    test.Canvas().Focus(entry)
```
Same result...

Reply:
This is my implementation:

```go
package widget

import (
	"log"

	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/data/binding"
	"fyne.io/fyne/v2/widget"
)

// CompletionEntry is an Entry with options displayed in a PopUpMenu.
type CompletionEntry struct {
	widget.Entry
	popupMenu     *widget.PopUp
	navigableList *navigableList
	items         []string
	pause         bool
}

// NewCompletionEntry creates a new CompletionEntry which creates a popup menu that responds to keystrokes to navigate through the items without losing the editing ability of the text input..
func NewCompletionEntry(options []string) *CompletionEntry {
	c := &CompletionEntry{items: options}
	c.ExtendBaseWidget(c)
	// c.Entry = widget.NewEntry()
	return c
}

// ShowCompletion displays the completion menu
func (c *CompletionEntry) ShowCompletion() {
	if c.pause {
		return
	}

	if c.navigableList == nil {
		c.navigableList = newNavigableMenu(c.items, &c.Entry, c.setTextFromMenu)
	}
	holder := fyne.CurrentApp().Driver().CanvasForObject(c)
	pos := fyne.CurrentApp().Driver().AbsolutePositionForObject(c)

	if c.popupMenu == nil {
		c.popupMenu = widget.NewPopUp(c.navigableList, holder)
	}
	c.popupMenu.ShowAtPosition(fyne.Position{X: pos.X, Y: pos.Y + c.Size().Height})
	c.popupMenu.Resize(fyne.NewSize(c.Entry.Size().Width, c.Size().Height*float32(len(c.items))))
	holder.Focus(c.navigableList)
}

// HideCompletion hides the completion menu.
func (c *CompletionEntry) HideCompletion() {
	if c.popupMenu != nil {
		c.popupMenu.Hide()
	}
}

// SetOptions set the completion list with itemList.
func (c *CompletionEntry) SetOptions(itemList []string) {
	c.items = itemList
	if c.navigableList != nil {
		c.navigableList.SetOptions(itemList)
	}
}

// Prevent the menu to open when the user validate value from the menu.
func (c *CompletionEntry) setTextFromMenu(s string) {
	c.pause = true
	c.Entry.SetText(s)
	c.Entry.Refresh()
	c.pause = false
	c.popupMenu.Hide()
}

type navigableList struct {
	*widget.List
	data            binding.StringList
	entry           *widget.Entry
	selected        int
	setTextFromMenu func(string)
	navigating      bool
}

func newNavigableMenu(items []string, entry *widget.Entry, setTextFromMenu func(string)) *navigableList {
	n := &navigableList{entry: entry, selected: -1, setTextFromMenu: setTextFromMenu}
	n.data = binding.BindStringList(&items)

	n.List = widget.NewListWithData(n.data, func() fyne.CanvasObject {
		return widget.NewLabel("")
	}, func(i binding.DataItem, o fyne.CanvasObject) {
		o.(*widget.Label).Bind(i.(binding.String))
	})

	n.OnSelected = func(id int) {
		if !n.navigating {
			item, _ := n.data.GetItem(id)
			val, _ := item.(binding.String).Get()
			n.entry.SetText(val)
		}
		n.navigating = false
	}

	return n
}

func (n *navigableList) SetOptions(items []string) {
	n.Unselect(n.selected)
	n.data.Set(items)
	n.selected = -1
}

func (n *navigableList) TypedKey(event *fyne.KeyEvent) {
	n.entry.TypedKey(event)
	switch event.Name {
	case fyne.KeyDown:
		if n.selected < n.data.Length()-1 {
			n.selected++
		}
		n.navigating = true
		n.List.Select(n.selected)

	case fyne.KeyUp:
		if n.selected > 0 {
			n.selected--
		} else {
			n.Unselect(n.selected)
			n.selected = -1
		}
		n.navigating = true
		n.List.Select(n.selected)
	case fyne.KeyReturn:
		n.navigating = false
		n.OnSelected(n.selected)
	}
}
func (n *navigableList) TypedRune(r rune) {
	n.entry.TypedRune(r)
}
func (n *navigableList) FocusGained() {
	log.Println("Focused !")
}
func (n *navigableList) FocusLost() {}

```

Reply:
And dont take care of names, I will change it later before to refresh the PR

Reply:
> Yes, this should be fixed, basically the problem is that each overlay has its own focus manager.

They do indeed, but calling `canvas.Focus` will find the correct manager.

Reply:
> This is not the wanted behavior because the cursor disappears from the entry

This is going to be the main challenge, because the "top most" item must have focus (so you can navigate the list) but you want it to look like the entry has focus. This may be possible by extending the entry to encourage it to look focused despite not being. However it may not be possible - it may require a little brainstorming with the team as well because I think the right balance is not obvious.

Reply:
> They do indeed, but calling `canvas.Focus` will find the correct manager.

This is what I did but the focus is not set. Is this a problem with Fyne ?

Reply:
Ho... It seems that win.Canvas().Focused() returned now the entry... So... Maybe...

Reply:
> This is what I did but the focus is not set. Is this a problem with Fyne ?

I don't know without looking at some code in detail, but I suspect the code should work or things like the Form dialog would not work. Maybe you did not add the overlay before calling focus like @fpabl0 suggested?

Reply:
No worry, it's now OK. I don't know what happens earlier but the tests are now OK. Excepting that I don't like the "list" usage (see my last commits)

Reply:
> This is going to be the main challenge, because the "top most" item must have focus (so you can navigate the list) but you want it to look like the entry has focus.

@andydotxyz it should not be hard, as they do not share the focus manager, so the entry will be focused in content focus manager, and the navigateList will be focused in the top most overlay focus manager.